_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,353 @@
+
+```scala
+import shapeless.=:!=
+  def foo[Ole](a: Ole)(implicit ev: =:!=[Ole, Int]): A = a
+
+  foo[String]("E.Near rocks.")   /* Compiles */
+  foo[Int](5)                    /* Does not compile */
+  foo[Boolean](true)             /* Compiles */
+```
+
+Briefly, if <code>Ole</code> and Int are different, the compiler will be able to (always) find the required implicit and succeed the compilation. If they are the same, something else happens and a compilation error is raised. How exactly using the syntax (implicit ev: =:!=[Ole, Int]) achieves that behavior is beyond this post and my explanation skills; it should be accepted by you as the truth.
+
+This can however be used to tackle our problem:
+
+```scala
+import shapeless.=:!=
+case class Reservation[SeatInfo <: Seat](
+  userTicket: Ticket[SeatInfo],
+  guestTicket: Ticket[SeatInfo]
+)(implicit ev: SeatInfo =:!= Seat)
+```
+
+The implicit evidence ev will demand the compiler to find proof that the <code>SeatInfo</code>, whatever type it might be, must not be type <code>Seat</code>. If it cannot, it throws a compilation error.  With figure 1 as guidance, because abstract type <code>SeatInfo</code> has as upper bound the concrete type <code>Seat</code>, and because of the generalized type constraint, <code>SeatInfo</code> must not be <code>Seat</code>, then it must surely only be one of <code>Standing</code>, <code>Viptribune</code> or <code>NormalSeat</code>. Problem solved. So, in contrast with the other 2 solutions so far seen, an attempt to instantiate a reservation with unsound parameters is caught during compilation:
+
+```scala
+val standing = Standing
+val normalSeated = NormalSeat(4, 5)
+val tribuneSeat = VipTribuneSeat("east")
+
+//  Compiles
+val bar = Reservation(
+  Ticket("id1", 123, normalSeated),
+  Ticket("id2", 456, normalSeated)
+)
+
+//  Will never compile
+val foo = Reservation(
+  Ticket("id1", 123,  standing),
+  Ticket("id2", 456, tribuneSeat)
+)
+
+//  Will never compile
+val baz = Reservation(
+  Ticket("id1", 123, normalSeated),
+  Ticket("id2", 456, tribuneSeat)
+)
+```
+
+When it does not compile we get the error:  
+ 
+```
+ambiguous implicit values:
+[error]  both method neqAmbig1 in package shapeless of type [A]=> A =:!= A
+[error]  and method neqAmbig2 in package shapeless of type [A]=> A =:!= A
+[error]  match expected type com.github.cmhteixeira.invariants.Seat =:!= com.github.cmhteixeira.invariants.Seat
+[error] val foo = Reservation(
+[error]                           ^
+```
+
+The error is not very helpful, but one must live with it. (footnote: answer by Régis Jean-Gilles on https://stackoverflow.com/questions/6909053/enforce-type-difference)  
+
+This approach, similarly to the 2 other approaches, suffers from an inconvenient. The type SeatInfo on a given reservation is lost at run time due to type erasure. We cannot pattern match on a given reservation as on the following example:  
```

Comment:
suffers from an inconvenient **____**

missing a noun, `inconvenient` should describe something

Reply:
inconvenient can actually be used as a noun as well.

Reply:
agree, as a noun it needs different ending `inconvenience`

Reply:
It seems you are right.  I will change.
Funny, I was very sure that `inconvenient` could also be used as a noun.